File: ecommerce/lib/models/product.dart
import 'package:ecommerce/models/ulasan.dart';
import 'package:ecommerce/models/variants.dart';

class Product {
  final String? id;
  final String? name;
  final String? sellerId;
  final int? price;
  final List<String>? category;
  final int? stockQuantity;
  final int? salesLastWeek;
  final int? totalUnitsSold;
  final String? sellerName;
  final String? imageUrl;
  final List<Ulasan>? reviews;
  final List<Variants>? variants;

  Product(
      {this.id,
      this.name,
      this.sellerId,
      this.price,
      this.category,
      this.stockQuantity,
      this.salesLastWeek,
      this.totalUnitsSold,
      this.sellerName,
      this.reviews,
      this.variants,
      this.imageUrl});

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['_id'],
      name: json['name'],
      sellerId: json['sellerId'],
      price: json['price'],
      category:
          (json['category'] as List<dynamic>).map((e) => e as String).toList(),
      stockQuantity: json['stockQuantity'],
      totalUnitsSold: json['totalUnitsSold'],
      salesLastWeek: json['salesLastWeek'],
      sellerName: json['sellerName'],
      imageUrl: json['imageUrl'],
      reviews: json['reviews'] != null
          ? (json['reviews'] as List<dynamic>)
              .map((reviewJson) =>
                  Ulasan.fromJson(reviewJson as Map<String, dynamic>))
              .toList()
          : null,
      variants: (json['variants'] as List<dynamic>?)
          ?.map((variantJson) =>
              Variants.fromJson(variantJson as Map<String, dynamic>))
          .toList(),
    );
  }

  factory Product.fromMap(Map<String, dynamic> map) {
    return Product(
      id: map['id'],
      name: map['name'],
      price: map['price'],
      sellerId: map['sellerId'],
      category:
          (map['category'] as List<dynamic>).map((e) => e as String).toList(),
      stockQuantity: map['stockQuantity'],
      totalUnitsSold: map['totalUnitsSold'],
      salesLastWeek: map['salesLastWeek'],
      sellerName: map['sellerName'],
      imageUrl: map['imageUrl'],
      reviews: map['reviews'] != null
          ? (map['reviews'] as List<dynamic>)
              .map((reviewmap) =>
                  Ulasan.fromJson(reviewmap as Map<String, dynamic>))
              .toList()
          : null,
      variants: (map['variants'] as List<dynamic>?)
          ?.map((variantJson) =>
              Variants.fromJson(variantJson as Map<String, dynamic>))
          .toList(),
    );
  }
}

File: server/models/faker-data/product.js
const Products = require("../product");
const sellerIds = require("./dummy/sellerIds");

const generateRandomProduct = () => {
  return new Products({
    name: faker.commerce.productName(),
    sellerId: sellerIds[Math.floor(Math.random() * sellerIds.length)],
    price: faker.commerce.price(),
    category: [faker.commerce.department(), faker.commerce.department()],
    stockQuantity: faker.datatype.number(100),
    salesLastWeek: faker.datatype.number(100),
    totalUnitsSold: faker.datatype.number(100),
    imageUrl: faker.image.abstract(),
    diskon: faker.datatype.number({ min: 0, max: 50 }), // Menghasilkan diskon antara 0 dan 50 persen
    ulasan: faker.lorem.paragraph(), // Menghasilkan ulasan produk yang lebih realistis
    variants: [
      {
        name: faker.commerce.productName(),
        price: faker.commerce.price(),
        imageUrl: faker.image.abstract(),
      },
      {
        name: faker.commerce.productName(),
        price: faker.commerce.price(),
        imageUrl: faker.image.abstract(),
      },
    ],
  });
};

const createFakeProducts = async (num) => {
  await mongoose.connect(
    "mongodb+srv://motherbloodss:XKFofTN9qGntgqbo@cluster0.ejyrmvc.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0",
    {
      useNewUrlParser: true,
      useUnifiedTopology: true,
    }
  );

  const products = [];
  for (let i = 0; i < num; i++) {
    products.push(await generateRandomProduct());
  }

  await Products.insertMany(products);
  console.log(`${num} fake products inserted successfully`);
  await mongoose.disconnect();
};

createFakeProducts(100);
